### PR TITLE
Challenge ACME

### PR DESCRIPTION
**Pull request by nuno-morais**
# Description
ECME Stores is a challenge for Software Angels. The goal of this challenge is to aggregate data from multiple sources and provide a GUI to list and edit some details.

## Decisions
As some requirements were not well defined, I've made some decisions about this challenge. However, in the event of scaling or requirements changing, it may be necessary to reevaluate some of these decisions, especially regarding resilience.

- This challenge asks for syncing data from multiple sources. In this case, I've decided to keep it simple, storing all details denormalized on the database.

- Since one of the requirements was to use Vue.js and Kotlin, the frameworks chosen to implement this project was SpringBoot + Nuxtjs. Given I have some knowledge related to SpringBoot, I thought it could be interesting in using it here.

- I've decided to take some risks and incorporate an experimental technology. To take snapshots from the database, I used Reactor, and in that case I used it to provide a way to throttling the number of queries to DB.

- The backend (BE) app was split in two:
  * Consumer: it is responsible to sync the data between certain time intervals;
  * Web: it is responsible to run the API.

* To implement the tests, I chose to use Spek, which is an awesome tool to give some proximity with BDD.

### Next Steps

**BE:**
- Improve the algorithm to search on the DB or use something that already solves the problem for us (full search engine);
- Organize the build.gradle in submodules (remove some dependencies from root);
- Introduce WireMock to do some stubs for integration tests (gateways);
- Split unit, integration, and acceptance tests properly;
- Add observability (e.g. New Relic, Bugsnag);
- Use Flyway instead of JPA Generator;
- Improve exception handlers.

**FE:**
- Use store and actions;
- Unify layout and styles;
- Add tests.

Notes:
- Sorry for committing the `API KEY` and `IP`

### CODE REVIEW

**Comment by voxivoid** [comment]
Hello @nuno-morais!

I'm André Domingues and I'm a developer within the company.

I would like to thank you the time you spent doing the challenge.

We will try to give you as much feedback as possible! 😄

File: readme.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+- Background Sync
+```shell
+docker-compose build consumer && docker-compose run consumer
+```
+
+- Frontend (FE)
+```shell
+yarn install
+yarn dev
+```
```
Comment (voxivoid) [inline readme.md:32]: is it possible to have a single docker-compose for both projects?

Comment (nuno-morais) [reply readme.md:32]: Yes, I can do it. 

BTW, I didn't use docker compose to serve the FE side. Do you want to up it using Docker?

**Comment by nuno-morais** [comment]
> e time you spent doing the chal

Hello,
Nice to meet you :) 

Thanks 🙏

**Comment by nuno-morais** [comment]
Sorry, **TIL** that I can't rebase the root commit from the main branch. It was intent because I had the spring-boot bootstrap there however, I didn't know that the current PR was closed automatically. That said, I've opened a new one [here](https://github.com/nuno-morais/acme/pull/2).

**Pull request closed without merging**